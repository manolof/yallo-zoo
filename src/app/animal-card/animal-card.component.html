<ng-container *ngIf="animal">
	<a
		[routerLink]="['/animals', animal.id]"
		class="block bg-white shadow-md hover:shadow-xl rounded-lg transition-shadow"
	>
		<figure>
			<div class="relative pb-40">
				<img
					class="absolute inset-0 h-full w-full object-cover rounded-t-lg"
					[src]="animal.avatar"
					[alt]="animal.name"
				/>
			</div>
			<figcaption class="p-4">
				<h2 class="mt-2 mb-2 font-bold">
					{{ animal.name }}
				</h2>
				<p class="text-sm">
					{{ animal.type }}
				</p>
			</figcaption>
		</figure>
	</a>
</ng-container>
