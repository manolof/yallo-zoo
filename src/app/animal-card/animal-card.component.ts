import { Component, Input, ChangeDetectionStrategy } from '@angular/core';

import { Animal } from '../animal';

@Component({
	selector: 'app-animal-card',
	templateUrl: './animal-card.component.html',
	changeDetection: ChangeDetectionStrategy.OnPush,
})
export class AnimalCardComponent {
	@Input() public animal!: Animal;
}
