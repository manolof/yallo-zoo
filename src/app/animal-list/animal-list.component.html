<div class="my-8">
	<ng-container *ngIf="animals$ | async as animals; else loading">
		<div class="grid gap-8 lg:gap-16 lg:grid-cols-3 md:grid-cols-2 px-8">
			<ng-container *ngFor="let animal of animals">
				<app-animal-card [animal]="animal"></app-animal-card>
			</ng-container>
		</div>

		<div class="flex justify-center my-8 lg:my-16">
			<button
				(click)="updateAnimals()"
				class="bg-teal-400 hover:bg-teal-500 text-white font-medium py-2 px-16 rounded-full transition-colors"
			>
				Update
			</button>
		</div>
	</ng-container>
	<ng-template #loading>
		<div class="px-8">Loading...</div>
	</ng-template>
</div>
