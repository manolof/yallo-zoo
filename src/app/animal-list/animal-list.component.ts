import {
	ChangeDetectionStrategy,
	Component,
	inject,
	OnInit,
} from '@angular/core';

import { HttpService } from '../http.service';
import { Animal } from '../animal';
import { Observable } from 'rxjs';

@Component({
	selector: 'app-animal-list',
	templateUrl: './animal-list.component.html',
	changeDetection: ChangeDetectionStrategy.OnPush,
})
export class AnimalListComponent implements OnInit {
	public animals$!: Observable<Animal[]>;

	private httpService = inject(HttpService);

	public ngOnInit() {
		this.animals$ = this.httpService.getAnimals();
	}

	public updateAnimals() {
		this.animals$ = this.httpService.getAnimals(true);
	}
}
