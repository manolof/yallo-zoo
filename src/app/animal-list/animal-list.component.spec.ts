import { ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { of } from 'rxjs';
import { AnimalCardComponent } from '../animal-card/animal-card.component';

import { HttpService } from '../http.service';
import { AnimalListComponent } from './animal-list.component';

describe('AnimalListComponent', () => {
	let component: AnimalListComponent;
	let fixture: ComponentFixture<AnimalListComponent>;

	const httpServiceMock = {
		getAnimals: jest.fn(),
	};

	const animals = [
		{ id: 1, name: 'Tiger', type: 'Mammal', avatar: 'tiger.jpg' },
		{ id: 2, name: 'Lion', type: 'Mammal', avatar: 'lion.jpg' },
	];

	beforeEach(async () => {
		await TestBed.configureTestingModule({
			imports: [RouterTestingModule],
			declarations: [AnimalListComponent, AnimalCardComponent],
			providers: [{ provide: HttpService, useValue: httpServiceMock }],
		}).compileComponents();

		fixture = TestBed.createComponent(AnimalListComponent);
		component = fixture.componentInstance;
	});

	it('should create', () => {
		expect(component).toBeTruthy();
	});

	it('should show loading text if still retrieving data', () => {
		httpServiceMock.getAnimals.mockReturnValue(undefined);
		component.ngOnInit();
		fixture.detectChanges();
		expect(fixture).toMatchSnapshot();
	});

	it('should fetch and render the animals list from the API', () => {
		httpServiceMock.getAnimals.mockReturnValue(of(animals));
		component.ngOnInit();
		fixture.detectChanges();
		expect(fixture).toMatchSnapshot();
	});

	it('should refresh the animals list', () => {
		httpServiceMock.getAnimals.mockReturnValue(of(animals));
		component.updateAnimals();
		fixture.detectChanges();
		expect(fixture).toMatchSnapshot();
	});
});
