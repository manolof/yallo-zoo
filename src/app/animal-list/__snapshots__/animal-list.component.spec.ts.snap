// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimalListComponent should fetch and render the animals list from the API 1`] = `
<app-animal-list
  __ngContext__={[Function Number]}
  animals$={[Function Observable]}
  httpService={[Function Object]}
>
  <div
    class="my-8"
  >
    <div
      class="grid gap-8 lg:gap-16 lg:grid-cols-3 md:grid-cols-2 px-8"
    >
      <app-animal-card>
        <a
          class="block bg-white shadow-md hover:shadow-xl rounded-lg transition-shadow"
          href="/animals/1"
        >
          <figure>
            <div
              class="relative pb-40"
            >
              <img
                alt="Tiger"
                class="absolute inset-0 h-full w-full object-cover rounded-t-lg"
                src="tiger.jpg"
              />
            </div>
            <figcaption
              class="p-4"
            >
              <h2
                class="mt-2 mb-2 font-bold"
              >
                 Tiger 
              </h2>
              <p
                class="text-sm"
              >
                 Mammal 
              </p>
            </figcaption>
          </figure>
        </a>
      </app-animal-card>
      <app-animal-card>
        <a
          class="block bg-white shadow-md hover:shadow-xl rounded-lg transition-shadow"
          href="/animals/2"
        >
          <figure>
            <div
              class="relative pb-40"
            >
              <img
                alt="Lion"
                class="absolute inset-0 h-full w-full object-cover rounded-t-lg"
                src="lion.jpg"
              />
            </div>
            <figcaption
              class="p-4"
            >
              <h2
                class="mt-2 mb-2 font-bold"
              >
                 Lion 
              </h2>
              <p
                class="text-sm"
              >
                 Mammal 
              </p>
            </figcaption>
          </figure>
        </a>
      </app-animal-card>
    </div>
    <div
      class="flex justify-center my-8 lg:my-16"
    >
      <button
        class="bg-teal-400 hover:bg-teal-500 text-white font-medium py-2 px-16 rounded-full transition-colors"
      >
         Update 
      </button>
    </div>
  </div>
</app-animal-list>
`;

exports[`AnimalListComponent should refresh the animals list 1`] = `
<app-animal-list
  __ngContext__={[Function Number]}
  animals$={[Function Observable]}
  httpService={[Function Object]}
>
  <div
    class="my-8"
  >
    <div
      class="grid gap-8 lg:gap-16 lg:grid-cols-3 md:grid-cols-2 px-8"
    >
      <app-animal-card>
        <a
          class="block bg-white shadow-md hover:shadow-xl rounded-lg transition-shadow"
          href="/animals/1"
        >
          <figure>
            <div
              class="relative pb-40"
            >
              <img
                alt="Tiger"
                class="absolute inset-0 h-full w-full object-cover rounded-t-lg"
                src="tiger.jpg"
              />
            </div>
            <figcaption
              class="p-4"
            >
              <h2
                class="mt-2 mb-2 font-bold"
              >
                 Tiger 
              </h2>
              <p
                class="text-sm"
              >
                 Mammal 
              </p>
            </figcaption>
          </figure>
        </a>
      </app-animal-card>
      <app-animal-card>
        <a
          class="block bg-white shadow-md hover:shadow-xl rounded-lg transition-shadow"
          href="/animals/2"
        >
          <figure>
            <div
              class="relative pb-40"
            >
              <img
                alt="Lion"
                class="absolute inset-0 h-full w-full object-cover rounded-t-lg"
                src="lion.jpg"
              />
            </div>
            <figcaption
              class="p-4"
            >
              <h2
                class="mt-2 mb-2 font-bold"
              >
                 Lion 
              </h2>
              <p
                class="text-sm"
              >
                 Mammal 
              </p>
            </figcaption>
          </figure>
        </a>
      </app-animal-card>
    </div>
    <div
      class="flex justify-center my-8 lg:my-16"
    >
      <button
        class="bg-teal-400 hover:bg-teal-500 text-white font-medium py-2 px-16 rounded-full transition-colors"
      >
         Update 
      </button>
    </div>
  </div>
</app-animal-list>
`;

exports[`AnimalListComponent should show loading text if still retrieving data 1`] = `
<app-animal-list
  __ngContext__={[Function Number]}
  animals$="undefined"
  httpService={[Function Object]}
>
  <div
    class="my-8"
  >
    <div
      class="px-8"
    >
      Loading...
    </div>
  </div>
</app-animal-list>
`;
